### ISSUE TITLE
min_similarity & max_distance does not work in levsim

### ISSUE THREAD

**Issue by BitMindLab** [RaRe-Technologies/gensim#2541]
#### Expecting

As doc says, `max_distance` is a more efficient way, more quickly.
https://github.com/RaRe-Technologies/gensim/blob/369cc638225a2080faec25c4e2f6448d31e0492b/gensim/similarities/levenshtein.py#L32-L37

#### Problem description

However, `max_distance` does not speed up the process

https://github.com/RaRe-Technologies/gensim/blob/369cc638225a2080faec25c4e2f6448d31e0492b/gensim/similarities/levenshtein.py#L48-L51

**Comment by piskvorky** [comment]
@BitMindLab please fill in the issue template fully: what you're trying, what you're expecting, what you're seeing instead.

**Comment by gojomo** [comment]
The doc-comment clearly does not match the code behavior, which gains no efficiency benefit from the specification of `max_distance` - it just sometimes returns a different value.

**Comment by piskvorky** [comment]
Whether the code does something more efficient or not is up to the code. I proposed (much) faster early-out variants in the past, but IIRC @Witiko didn't think it was worth it.

The only problem I see is that the documentation is unclear whether supplying `max_distance` **must** speed up computation, or simply **may** speed up computation. The intent is **may** (and with the current implementation, supplying the parameter has no real effect, but that may change in the future).

@Witiko @gojomo @BitMindLab @mpenkov can you clarify the docstring wording? Or even better, implement the early-out stop when `max_distance` is know more intelligently…

**Comment by Witiko** [comment]
@piskvorky As discussed in https://github.com/RaRe-Technologies/gensim/pull/2016#discussion_r202326763, replacing the `levsim` function with a stub (i.e. Levenshtein distance is not even computed) leads to less than 2× speed improvement. My conclusion was that rather than optimize the pointwise `levsim` function, which will lead to virtually no speed-up, we should be computing the Levenshtein distance between all pairs of words in bulk.

**Comment by piskvorky** [comment]
That's the one, thanks for digging it up :)

I assume "2×" also depends on the particular inputs, no? I.e. the longer the strings => more costly to compute exact Levenshtein (especially compared to a low `max_distance` threshold).

Replacing naive pair-wise computations by a better algo sounds good too. Which algo do you have in mind? How would that work?

**Comment by Witiko** [comment]
The average time complexity of computing the edit distance for all pairs using a pairwise algorithm is O(M²N²), where M is dictionary size and N is average word length. In fact, a bit tighter bound exists with respect to N, since a subquadratic algorithm was recently discovered, see [here](https://arxiv.org/abs/1810.03664). However, since N is constant for a language, the actual average time complexity is O(M²), i.e. there is virtually no difference between computing pointwise edit distance and calling a stub, as seen in the measurements.

I am currently unaware of any algorithm for computing the edit distance in bulk, but I am looking. We can gain a little speed-up by optimizing the python-Levenshtein library, but the improvements will be insignificant (for English).

**Comment by piskvorky** [comment]
The improvements of early-out over using full Levenshtein distance are many-fold. I've seen 10x+, gains (again, depending on string length: the longer the inputs, the bigger the gain).

Finding a set of all words from a fixed (static) lexicon that are within a given edit distance from a dynamic query word is a well-studied problem. Trie and automata are the typical solution.

I've done this multiple times before (spelling correction etc), it should be fun to optimize this Gensim module too. But probably not any time soon, no spare time now, no itch to self-scratch.

**Comment by Witiko** [comment]
> If the "normal" words you tried on all equally small, that's great, but is it true for all supported use-cases / languages?

In the benchmark, we used the dictionary of the Google News dataset (English). We would see a larger speed-up for languages with longer words such as German.

> The API claims it's looking for `Get most similar terms for a given term.`:

This is true. In the `SparseTermSimilarityMatrix` class, we eventually call `most_similar` for every term, which is equivalent to computing edit distance between all word pairs in our current implementation. However, since we are only interested in the k nearest neighbors, our problem is easier.

> Finding a small set of all words from a **fixed static lexicon** that are within a given distance from a **dynamic query word** is a well-studied problem. Trie and automata are the typical solution. It can be done extremely fast.
>
> I've done this multiple times before myself (for spelling correction, incl. with custom edit distance fnc to give different edit weights to typos, character accent marks "ř" vs "r", keyboard layouts "y" vs "z" etc). It should be fun to optimize this Gensim module too. But probably not any time soon, no spare time now, no itch to self-scratch.

Here are some references:

1. Nick J., 2007. Nick's Blog. **Damn Cool Algorithms, Part 1: BK-Trees**. [[URL](http://blog.notdot.net/2007/4/Damn-Cool-Algorithms-Part-1-BK-Trees)]
2. Nick J., 2010. Nick's Blog. **Damn Cool Algorithms: Levenshtein Automata**. [[URL](http://blog.notdot.net/2010/07/Damn-Cool-Algorithms-Levenshtein-Automata)]
3. Boytsov, L., 2011. Journal of Experimental Algorithmics (JEA). **Indexing methods for approximate dictionary searching: Comparative analysis**. [[URL](http://searchivarius.org/personal/leonid-boytsovs-publications#jea2011)]

Regrettably, I will have no time to work on this during July and August.

**Comment by Witiko** [comment]
@piskvorky Two years later and I came across a simple solution, which should reduce the complexity of our kNN algorithm over the Levenshtein distance from O(M²) to O(M log M), where M is the number of indexed words. The solution would use [the ball tree from scikit-learn](https://github.com/scikit-learn/scikit-learn/discussions/20097#discussioncomment-742090), which is a structure for indexing metric spaces, and which reduces the number of comparisons from M to log M when the tree is balanced compared to our current naïve brute-force approach.

There exist better algorithms specifically for kNN over the Levenshtein distance, but their implementations in Python are elusive. The ball tree is implemented in a core Python library, which is also a Gensim dependency, and the use is straightforward. We may wish to explore further opportunities for a speed up using approximate kNN in the future, but this seems as a major improvement over the current situation.

**Comment by piskvorky** [comment]
Sure, if that works. I vaguely remember trying ball-tree and it didn't work well, but that was in a different context. I'd have to "load up" this thread into my memory to remember what it's about, but I'm sure you already did that :-)

Let me know when you get some numbers!

**Comment by Witiko** [comment]
Today, I played around with the ball tree a bit. However, I was disappointed to discover that sklearn's Ball tree assumes Euclidean vectors and uses centroids to represent intermediate nodes, which rules out non-vector metric spaces such as strings over the Levenshtein distance.

I discovered a number of metric tree implementations in Python. However, I only managed to double the speed, which is arguably insufficient to make kNN search over the Levenshtein distance practically useful. I placed measurements to #3146.

**Comment by piskvorky** [comment]
Pity. I don't remember the particulars of the problem you're solving here, but searching for "similar strings", where similar = "Damerau-Levenshtein within a given distance", can be done extremely fast, using trie and DAWG automata. I see that as a more viable approach than fancy clustering over metric spaces.

**Comment by Witiko** [comment]
I was curious how much speed we could gain with metric indexing, which is a nice general technique. I tried the DAWG automata as implemented by [lexpy][1]: we get 2x speed-up when finding strings within edit distance 2 (i.e. roughly the same as VT-Tree) and up to 10x speed-up when finding strings within edit distance 1. We would need ca 100x speed-up to be rougly in the same ballpark as vector space kNN search of word embeddings.

 [1]: https://pypi.org/project/lexpy/

**Comment by piskvorky** [comment]
10x speed-up compared to brute-forcing each comparison `query_string x lexicon_string`? That doesn't sound right.

I'd expect the speed up to depend on the number of answers = words within the lexicon that are actually within distance N. Any fixed number is suspicious. We're probably talking about different algos here.

**Comment by Witiko** [comment]
This speed-up is with regards to the small benchmark I set up at https://github.com/RaRe-Technologies/gensim/pull/3146.

**Comment by piskvorky** [comment]
OK I'll check #3146, to see how the overall problem is formulated, what is being measured.

**Comment by Witiko** [comment]
I am getting better results with the bigger [text8][2] corpus (100MB):

- ca 500 it/s with `WordEmbeddingSimilarityIndex`
- ca 0.8 it/s with the current brute-force kNN
- ca 1.6 it/s with VP-Tree kNN
- ca 8 it/s with DAWG searching for words within edit distance 2
- ca 80 it/s with DAWG searching for words within edit distance 1

This is already arguably quite nice, although still way too slow.

 [2]: http://mattmahoney.net/dc/textdata

**Comment by piskvorky** [comment]
See my comment there – I got several times faster results by using an automaton, in pure Python.

**Comment by piskvorky** [comment]
For future reference – if we only look for `k=1` (not larger `k`), we can probably bypass automata altogether, and generate the candidate words directly instead, to look up in the dictionary exactly.

This should be pretty fast; I just tried [TinyFastSS](https://github.com/fujimotos/TinyFastSS) and on text8 (253,854 terms in the dictionary) I'm getting ~700 `max_dist=1` queries per second. Pure Python, a single file of a few hundred lines of simple code.
